shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,313 @@
+     * was triggered by a done task from one of its tasks. This type works with the NUClear reaction system by providing
+     * a dereference operator so that the users can either choose to receive a TaskInfo<T> object or a const T& object
+     * in their callback if they don't care about the extra information the director provides.
+     *
+     * This object also ensures that the task will not run when the Director has no information for this Provider.
+     * This can be caused when it should not be running but was triggered by another bind statement (e.g. Every).
+     * In this case it has a bool operator that returns false which will stop the task from running.
+     *
+     * @tparam T the type of task that is stored in this task_info object
+     */
+    template <typename T>
+    struct TaskInfo {
+
+        /**
+         * Default constructor, as data will be null NUClear will be told not to execute this task.
+         */
+        TaskInfo() {}
+
+        /**
+         * Construct a new TaskInfo object for the provided task data
+         *
+         * @param data_ the data for this task
+         */
+        TaskInfo(std::shared_ptr<const T> data_) : data(data_) {}
+
+        /**
+         * Having a dereference operator allows NUClear to provide either the TaskInfo<T> object or a const T& object
+         *
+         * @return the data that is contained within this TaskInfo object
+         */
+        const T& operator*() {
+            return *data;
+        }
+
+        /**
+         * This will control if NUClear should run the task
+         *
+         * @return true     there is data, NUClear should run the task
+         * @return false    there is no data, NUClear should not run the task
+         */
+        operator bool() {
+            return data != nullptr;
+        }
+
+        /// The task data, this will be all that is returned if the user asks for a const T& rather than a TaskInfo
+        std::shared_ptr<const T> data;
+    };
+
+
+    /**
+     * This type is used as a base extension type for the different Provider DSL keywords (Provides, Entering, Leaving)
+     * to handle their common code.
+     *
+     * @tparam T        The type that this Provider services
+     * @tparam action   The action type that this Provider is
+     */
+    template <typename T, commands::ProviderClassification action>
+    struct ProviderBase {
+
+        /**
+         * Binds a Provider object, sending it to the Director so it can control its execution.
+         *
+         * @tparam DSL the NUClear DSL for the on statement
+         *
+         * @param reaction the reaction object that we are binding the Provider to
+         */
+        template <typename DSL>
+        static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+            // Tell the Director
+            reaction->reactor.powerplant.emit(
```

Comment:
Can `reaction` ever be `nullptr` here?

Reply:
NUClear would explode if it were possible :P